### PR TITLE
Local reactive template state

### PR DESCRIPTION
**Pull request by stubailo**
You can now use `Template.instance().state` to store local reactive data for your template. There is also a `Template.state` built-in helper that you can use to access it without defining a new helper.

to do:
1. Add serialization hooks for migration
2. Refactor reactive dict to not always serialize things

### CODE REVIEW

**Comment by stubailo** [comment]
Still need to add tests for migration.

**Comment by stubailo** [comment]
Also add test to check that the keyword `Template.state` is indeed reactive when you change the argument and when you set something on state.

File: packages/spacebars-compiler/codegen.js [hunk 1]
```diff
@@ -27,7 +27,9 @@ var builtInTemplateMacros = {
   // Confusingly, this makes `{{> Template.dynamic}}` an alias
   // for `{{> __dynamic}}`, where "__dynamic" is the template that
   // implements the dynamic template feature.
-  'dynamic': 'Template.__dynamic'
+  'dynamic': 'Template.__dynamic',
+
+  'state': '_.bind(view.templateInstance().state.get, view.templateInstance().state)'
```
Comment (avital) [inline packages/spacebars-compiler/codegen.js]: The resulting JS that gets compiled from templates should look clean. If this dance is what we need, define a utility function in the spacebars package (where the runtime for Spacebars compiled templates live).

Comment (dgreensp) [inline packages/spacebars-compiler/codegen.js]: Or `'state': 'Template.__state'`

Comment (stubailo) [inline packages/spacebars-compiler/codegen.js]: I like this idea, except wouldn't I need to pass the view in as an argument?

**Comment by dgreensp** [comment]
I haven't finished reading this yet, but what is the template syntax?  `{{Template.state "foo"}}`?

We don't have to block on having an answer for this, but the obvious question is how to make that better so that you can use state in, say, an argument to a helper.  Something closer to `state.foo`.

File: packages/spacebars-compiler/codegen.js (hunk 1 above)
Comment (dgreensp) [inline packages/spacebars-compiler/codegen.js]: Hmm, don't we have a dynamic variable for that?  `Template.instance()`

**Comment by stubailo** [comment]
I don't think you would use state in an argument to a helper usually. I think `Template.state "foo"` is the best we can do until we improve scoping mechanics.

File: packages/spacebars-compiler/codegen.js (hunk 1 above)
Comment (stubailo) [inline packages/spacebars-compiler/codegen.js]: Good idea!

**Comment by dalerka** [comment]
Hey guys! My apologies if this is not relevant here (looking forward for [changes described](https://medium.com/@stubailo/upcoming-features-in-meteor-blaze-b537cf9c09e9) by @stubailo): 

To keep the Meteor API as succinct as possible (easier for learning and using) perhaps it is better to have one "name" for some functionality instead of two, i.e to have `this.state` and `Template.state` to *set* and *get* some values instead of having another `this.vars` (which is less semantic than `this.state` IMHO):
```javascript
this.vars.setDefault(“selectedDay”, new Date());
... 
{{Template.state.selectedDay}}
```

**Comment by stubailo** [comment]
@dalerka the inconsistency was a typo on my part in the blog post. After talking a little with @debergalis we are currently thinking of calling this feature "vars" as opposed to "state" because "state" seems too jargon-ey. Do you have any ideas for other names that will be easy to understand for new Meteor developers?

**Comment by dalerka** [comment]
@stubailo ok, thanks. If the anticipated API ("vars") is intended to help with managing the Template's *state of variables* then `vars` seems simple and easy to get used to. But if you guys have some ideas that this API will serve other purposes beyond stated above, then perhaps there could be another name. I hope it won't cause much confusion when talking about such *states* of Template as *rendered* or *created* etc. An alternative name that I can think of is `store`.

**Comment by lourd** [comment]
[This hackpad](https://meteor.hackpad.com/Lickable-Forms-and-Components-6CVspZsVwJY) that @gschmidt started a while back has a lot of ideas about API names, scoping, and general object-oriented patterns for Templates/Components that could be good to take a look at again

**Comment by stubailo** [comment]
Looks like there Geoff decided on `fields` where we currently have `vars` (the *.define syntax is a whole different thing that we are also potentially thinking about).

I'm not sure that `fields` is clearer than `vars` or `state`.

**Comment by dgreensp** [comment]
I brought up "fields" and Matt said it made him think of text fields, and I agree.

As a bit of trivia, the name "field" for reactive state originally dates back to when we were thinking of calling the current template instance Meteor.form!

**Comment by stubailo** [comment]
Wow, interesting. Looks like dynamic variables were all the rage back then.

**Comment by dburles** [comment]
+1 for naming it state. The conversations here (and elsewhere) I see it naturally referred to as the "Reactive state" of a template. vars on the other hand makes me think of ReactiveVar, which it is not. Especially if it's going to be clear in the docs that state is a ReactiveDict.

**Comment by stubailo** [comment]
@dburles personally I agree with you that `state` is more specific. I'm not really sure that `vars` is less jargon-y, especially since they are really just dictionary keys and not variables in their own right. I think it's right that we can just say "`this.state` is a reactive dictionary where templates can keep any local UI state, for example the current tab open or the state of a dropdown menu."

I wonder who we could ask to actually find out which one is better. React calls this thing state, so maybe it's better to be less confusing by using a similar word for a similar thing. @debergalis what do you think?

**Comment by debergalis** [comment]
okay, sure.

**Comment by mitar** [comment]
I would like to link to #3590. I think `ReactiveDict` needs some memory cleanup support before being useful here.

**Comment by Maxhodges** [comment]
would hate to see template.state repeat the same design mistake we see in ReactiveDict. Maybe now is a change to do it right?  

See https://github.com/meteor/meteor/issues/4100#issuecomment-104796748

**Comment by aadamsx** [comment]
Completely agree with @Maxhodges and @mitar on this, please give ReactveDict some love.

**Comment by stubailo** [comment]
Can we keep this thread on topic, please?

**Comment by aadamsx** [comment]
I understand this probably isn't the right spot to voice concerns/feedback about meteor implementations/features.  

Sorry about that.

To keep this a clean room, is there a spot for discussion with the MDG on topics such as ReactiveDict?  Or should someone like @Maxhodges or myself open a issue here to start a discussion?  (I assume feedback and discussion are welcomed by the MDG on any topic)

**Comment by stubailo** [comment]
Probably not going to happen soon, and this PR doesn't have that much work in it anyway.

**Comment by Sanjo** [comment]
@stubailo Can you give some insight why this feature is not merged? Would be interesting for me.

**Comment by stubailo** [comment]
I think I didn't actually add enough value to add a new feature - ideally, this wouldn't just be the same as doing:

```js
Template.x.onCreated(function () {
  this.state = new ReactiveDict();
});
```

I would have to build a lot more stuff to make it better, and I don't have time. So I'm explicitly abandoning this PR.

**Comment by mitar** [comment]
I use something similar with [Blaze Components](https://github.com/peerlibrary/meteor-blaze-components) and [reactive fields](https://github.com/peerlibrary/meteor-reactive-field).

**Comment by Sanjo** [comment]
Ok. I use ReavtiveVar right now for this. Works for me.

```js
var template = Template.x;

template.onCreated(function () {
  this.myReactiveStateVar1 = new ReactiveVar();
  this.myReactiveStateVar2 = new ReactiveVar();
});

template.helpers({
  myReactiveStateVar1: function () {
    return Template.instance().myReactiveStateVar1.get();
  ),

  myReactiveStateVar2: function () {
    return Template.instance().myReactiveStateVar2.get();
  )
});
```

**Comment by mitar** [comment]
Yea, in Blaze Components this becomes much simpler. :-) You do not have all this boilerplate:

```javascript
var ExampleComponent = BlazeComponent.extendComponent({
  onCreated: function () {
    this.myReactiveStateVar1 = new ReactiveField();
    this.myReactiveStateVar2 = new ReactiveField();
  }
}).register('ExampleComponent');
```

**Comment by rclai** [comment]
@mitar Yes, it is nice to be able to declare it and not have to write a helper to access it. We can just start using it in the template!

**Pull request closed without merging**